src/lib/field.ts

```diff
@@ -0,0 +1,1194 @@
+    return new Field(z);
+  }
+
+  /**
+   * Substract another "field-like" value from this {@link Field} element.
+   *
+   * @example
+   * ```ts
+   * const x = Field(3);
+   * const difference = x.sub(5);
+   *
+   * difference.assertEquals(Field(-2));
+   * ```
+   *
+   * **Warning**: This is a modular substraction in the pasta field.
+   *
+   * @example
+   * ```ts
+   * const x = Field(1);
+   * const difference = x.sub(Field(2));
+   *
+   * // If you try to print difference - `console.log(difference.toBigInt())` - you will realize that it prints a very big integer because this is modular arithmetic, and 1 - 2 circles around the field to become p - 1.
+   * // You can use the reverse operation of substraction (addition) to prove the difference is calculated correctly.
+   * difference.add(Field(2)).assertEquals(x);
+   * ```
+   *
+   * @param value - a "field-like" value to substract from the {@link Field}.
+   *
+   * @return A {@link Field} element equivalent to the modular difference of the two value.
+   */
+  sub(y: Field | bigint | number | string) {
+    return this.add(Field.from(y).neg());
+  }
+
+  /**
+   * Multiply another "field-like" value with this {@link Field} element.
+   *
+   * @example
+   * ```ts
+   * const x = Field(3);
+   * const product = x.mul(Field(5));
+   *
+   * product.assertEquals(Field(15));
+   * ```
+   *
+   * @param value - a "field-like" value to multiply with the {@link Field}.
+   *
+   * @return A {@link Field} element equivalent to the modular difference of the two value.
+   */
+  mul(y: Field | bigint | number | string): Field {
+    if (this.isConstant() && isConstant(y)) {
+      return new Field(Fp.mul(this.toBigInt(), toFp(y)));
+    }
+    // if one of the factors is constant, return Scale AST node
+    if (isConstant(y)) {
+      let z = Snarky.field.scale(this.value, Field.#toConst(y));
+      return new Field(z);
+    }
+    if (this.isConstant()) {
+      let z = Snarky.field.scale(y.value, this.value[1]);
+      return new Field(z);
+    }
+    // create a new witness for z = x*y
+    let z = Snarky.existsVar(() =>
+      FieldConst.fromBigint(Fp.mul(this.toBigInt(), toFp(y)))
+    );
+    // add a multiplication constraint
+    Snarky.field.assertMul(this.value, y.value, z);
+    return new Field(z);
+  }
+
+  /**
+   * [Modular inverse](https://en.wikipedia.org/wiki/Modular_multiplicative_inverse) of this {@link Field} element. Equivalent to 1 divided by this {@link Field}, in the sense of modular arithmetic.
+   *
+   * @example
+   * ```ts
+   * const someField = Field(42);
+   * const inverse = someField.inv();
+   * inverse.assertEquals(Field(1).div(example)); // This statement is always true regardless of the value of `someField`
+   * ```
+   *
+   * **Warning**: This is a modular inverse. See {@link div} method for more details.
+   *
+   * @return A {@link Field} element that is equivalent to one divided by this element.
+   */
+  inv() {
```

Comment:
Can we add in the comment that this throws on invert by zero? Should we have some warning about division by zero in-snark vars too?